### PR TITLE
issue #160 conditional read support

### PR DESCRIPTION
**Pull request by albertwang-ibm**
1. Added full support for conditional read:
(1) Support If-Modified-Since
      Support RFC_1123, RFC_850, ANSIC and touchstone date time formats.
(2) Support If-None-Match
      Support ETag value with or without " (and W/)
       e.g:  1, "1", W/1, W/"1" (the first format is used by TouchStone)
2. Added statement  of conditional read full support to server capability.
3. Added server integrations tests for conditional read.
4. Fixed the date time format of Last-Modified response header.
5. Updated last-modified response header test codes accordingly.

### CODE REVIEW

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java [hunk 1]
```diff
@@ -600,6 +618,30 @@ public Response conditionalDelete(@PathParam("type") String type) throws Excepti
             log.exiting(this.getClass().getName(), "delete(String,String)");
         }
     }
+    
+    
+    private long parseIfModifiedSince() {
```
Comment (prb112) [inline fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: We can probably use DateTimeHandler from Search's pattern and use a parseBest
Snippet

```
private static final DateTimeFormatter DATE_TIME_PARSER_FORMATTER =
            new DateTimeFormatterBuilder()
                    .appendPattern("yyyy")
                    .optionalStart()
                        .appendLiteral('-')
                        .appendPattern("MM")
                        .optionalStart()
                            .appendLiteral('-')
                            .appendPattern("dd")
                            .optionalStart()
                                .appendLiteral("T")
                                .optionalStart()
                                .appendPattern("HH")
                                .optionalStart()
                                    .appendLiteral(':')
                                    .appendPattern("mm")
                                    .optionalStart()
                                        .appendLiteral(':')
                                        .appendPattern("ss")
                                        .optionalStart()
                                            .appendFraction(ChronoField.MICRO_OF_SECOND, 0, 6, true)
                                        .optionalEnd()
                                    .optionalEnd()
                                .optionalEnd()
                                .optionalStart()
                                    .appendPattern("XXX")
                                .optionalEnd()
                            .optionalEnd()
                            .optionalEnd()
                        .optionalEnd()
                    .optionalEnd()
                    .toFormatter();
    // @formatter:on
```

then one call to 

```
return DATE_TIME_PARSER_FORMATTER.withResolverStyle(ResolverStyle.SMART).parseBest(value,
                    ZonedDateTime::from, LocalDateTime::from, LocalDate::from, YearMonth::from, Year::from);
```

Comment (albertwang-ibm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: en... seems the new pattern is much more complicated then using SimpleDateFormat, but let me try to change to use it ... :)

Comment (albertwang-ibm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: just noticed that it's mainly used for formater with optional elements, but in my case, we need 2 different formaters with the required different elements. so I would prefer to keep current implementation.

Comment (prb112) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: I don't think you need two formatters.  I think breaking this discussion out and talking with John may be best

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java [hunk 2]
```diff
@@ -162,6 +172,17 @@
     private static final String LOCAL_REF_PREFIX = "urn:";
     private static final String HEADERNAME_IF_NONE_EXIST = "If-None-Exist";
     private static final String HEADERNAME_PREFER = "Prefer";
+    private static final String HEADERNAME_IF_MODIFIED_SINCE = "If-Modified-Since";
+    private static final String HEADERNAME_IF_NONE_MATCH = "If-None-Match";
+
+    public static final DateTimeFormatter PARSER_FORMATTER = new DateTimeFormatterBuilder()
```
Comment (prb112) [inline fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: Maybe add a comment indicating what it supports? and format?

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java [hunk 3]
```diff
@@ -600,6 +621,28 @@ public Response conditionalDelete(@PathParam("type") String type) throws Excepti
...
     }
+    
+    
+    private long parseIfModifiedSince() {
+        long modifiedSince = -1;
+        try {
+            // Handle RFC_1123 and RFC_850 formats first. 
+            // e.g "Sun, 06 Nov 1994 08:49:37 GMT", "Sunday, 06-Nov-94 08:49:37 GMT", "Sunday, 06-Nov-1994 08:49:37 GMT"
+            // If 2 digits year is used, then means 1940 to 2039.
+            modifiedSince = httpServletRequest.getDateHeader(HEADERNAME_IF_MODIFIED_SINCE);
```
Comment (prb112) [inline fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: so it can be a long? or timestamp?

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java [hunk 4]
```diff
@@ -610,9 +653,50 @@ public Response read(@PathParam("type") String type, @PathParam("id") String id)
...
+            
+            String ifNoneMatch = httpHeaders.getHeaderString(HEADERNAME_IF_NONE_MATCH);
+            
             Resource resource = doRead(type, id, true, false, null, null, queryParameters);
-            ResponseBuilder response = Response.ok().entity(resource);
-            response = addHeaders(response, resource);
+            int version2Match = -1;
+            // Support ETag value with or without " (and W/)
+            // e.g:  1, "1", W/1, W/"1" (the first format is used by TouchStone)
+            if (ifNoneMatch != null && !ifNoneMatch.replaceAll("\"", "").replaceAll("W/", "").trim().isEmpty()) {
```
Comment (prb112) [inline fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: I think this part should be done one time, not twice as on 664/666

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java [hunk 5]
```diff
@@ -610,9 +653,50 @@ public Response read(@PathParam("type") String type, @PathParam("id") String id)
...
+            Instant modifiedTime2Compare = null;
+            if (modifiedSince > 0 ) {
+                modifiedTime2Compare = Instant.ofEpochMilli(modifiedSince);
+            }
+            
+            boolean isModifed = true;
+            // check if-not-match first
+            if (version2Match != -1) {
+                if (version2Match == Integer.parseInt(resource.getMeta().getVersionId().getValue())) {
+                    isModifed = false;
```
Comment (prb112) [inline fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: slight spelling ... Modified

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java (hunk 3 above)
Comment (albertwang-ibm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: long, the getDateHeader function returns EpochMilli. will add comment to clarify this.

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java (hunk 2 above)
Comment (albertwang-ibm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: good point, will add comment

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java (hunk 5 above)
Comment (albertwang-ibm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: ha, good catch!

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java (hunk 4 above)
Comment (albertwang-ibm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: I thought about this also, and because this requires another if block, so, I didn't. make sense to parse only once, will change.

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java (hunk 3 above)
Comment (JohnTimm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: I'm not sure that this pattern is exactly what we want. You have two optional segments and, technically, they could both be present and it would still work.

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java (hunk 4 above)
Comment (albertwang-ibm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: done

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java (hunk 5 above)
Comment (albertwang-ibm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: done

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java (hunk 3 above)
Comment (albertwang-ibm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: done

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java (hunk 2 above)
Comment (albertwang-ibm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: done

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java (hunk 3 above)
Comment (albertwang-ibm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: we want to support 2 different patterns with only one formatter, this is the only way so far which works in my tests. I don't need they work together, just need it works if any of them presents.  :( .   PARSER_FORMATTER.parse(httpHeaders.getHeaderString(HEADERNAME_IF_MODIFIED_SINCE), LocalDateTime::from).     and I require to parse as LocalDateTime, let me add test codes to find out what happens if both present, I would guess it will throw exception.

Comment (albertwang-ibm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: added more test codes to cover negative cases.

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /*
- * (C) Copyright IBM Corp. 2016,2019
+ * (C) Copyright IBM Corp. 2020
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: this was not right.  should have been 2016,2020

File: fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRResponseImpl.java [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /*
- * (C) Copyright IBM Corp. 2016,2019
+ * (C) Copyright IBM Corp. 2020
```
Comment (lmsurpre) [inline fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRResponseImpl.java]: not right.  should be 2016,2020
